### ISSUE TITLE
Print dialog has multiple issues on Windows 11 (no preview, no page numbers, window issues, etc.)

### ISSUE THREAD

**Issue by GabeS573** [musescore/MuseScore#16617]
### Issue type

UI bug

### Bug description

The Print dialog on Windows 11 has multiple issues.
- It does not show a preview of the document
- It does not show how many pages the document is
- If you choose to print a custom page range, the input isn't constrained to pages in the document (although only pages actually in the document will print)
- The Print dialog window is titled "Printing from Win32 Application. - Print" which is slightly janky
- The print dialog windows can't be accessed from the taskbar
- When MuseScore loses focus then gains it again, the main window goes in front of the still-open Print dialog window, sometimes making it look like the Print window closed (combined with the previous issue)

### Steps to reproduce

1. Open a score
2. Go to File > Print or press Ctrl + P
3. See error.

### Screenshots/Screen recordings

![image](https://user-images.githubusercontent.com/91428485/222297391-e848a038-1ef8-46f7-b44d-8f0155364335.png)

![image](https://user-images.githubusercontent.com/91428485/222297510-f23af865-4ff8-4414-9d8e-89cd7e216572.png)

https://user-images.githubusercontent.com/91428485/222299795-101bef67-4e61-4d9e-890a-8a69296d82e6.mp4

### MuseScore Version

MuseScore version (64-bit): 4.0.2-230551516, revision: github-musescore-musescore-d23f054

### Regression

Yes, this used to work in a previous version of Musescore 4.x

### Operating system

Windows 11 Home, version 22H2, OS build 22621.1265

### Additional context

**Context about "Regression":** This used to work in MuseScore 3 _and_ MuseScore 4, but no longer works in **EITHER** MuseScore 3 or 4, meaning this was likely caused by a Windows update.

I have no idea how much of this issue can actually be fixed on the MuseScore end of if some of this is just janky Windows behavior that can't be fixed by MuseScore.

**Comment by zacjansheski** [comment]
I don't seen the print preview functioning on Windows 11 previous versions of 4.  (Tried in 4.0, 4.01, 4.0.2)
MS3 does not show a preview either.

**Comment by Jojo-Schmitz** [comment]
It __is__ a regression, vs. Mu3 at least

And all these issue still exists.

On top: printing a score in continuous view doesn't work properly, Mu3 in that case temporarily switched to Page view, Mu4 aparently does not.

In essence the print functionallity is compeltly useless

**Comment by artoffugue333** [comment]
A Google search of this issue rendered the following workaround by using legacy Windows 10 . I'm not a coder so am wondering if this fix might be useful in fixing the bug in Musescore?

URL: https://www.google.com/search?q=windows+11+%22printing+from+win32+application%22+no+print+preview&oq=windows+11+%22printing+from+win32+application%22+no+print+preview&gs_lcrp=EgZjaHJvbWUyBggAEEUYOdIBCTI5ODM5ajBqN6gCCLACAQ&sourceid=chrome&ie=UTF-8

Excerpted steps to change registry in Win11:
1. Restoring the legacy print dialog.
2. Open the command window
3. Type reg add “HKCU\Software\Microsoft\Print\UnifiedPrintDialog” /v “PreferLegacyPrintDialog” /d 1 /t REG_DWORD /f

**Comment by Jojo-Schmitz** [comment]
It does indeed change the print dialog, but that still doesn't have a preview (but at least not an **__empty__** one). 
![image](https://github.com/musescore/MuseScore/assets/1786669/2f75997e-9c9f-420b-9adc-86c1360f9aa8)

**Comment by artoffugue333** [comment]
Good. I didnt

**Comment by Jojo-Schmitz** [comment]
Came up again in https://musescore.org/en/node/376295

**Comment by peelybird** [comment]
Yes still a weird issue for a composition tool to not support print preview in windows. Would love this to be fixed.

**Comment by MarcSabatella** [comment]
@peelybird - to be clear, MuseScore certainly has a print preview: the Publish tab.  It just doesn’t force you to go through an OS-level print dialog to get there, which depending on your OS may or may not work.

**Comment by cbjeukendrup** [comment]
This is a Qt limitation, so either someone should implement support for this at Qt, or we should implement printing for each OS separately using non-portable APIs.

**Comment by peelybird** [comment]
You lot are speedy responders, thank you! The preview tab you mention, @MarcSabatella, is useful for removing the clutter of the edit tools from the score, but when you actually press the print menu button from that very same tab, i am met with the same confusing "no print preview" issue in the print dialog. 
If I want to print in windows, i am "forced" to go through the print dialog, or am I wrong? And when that window doesn't confirm what I was seeing in either the publish tab or score tab, I am confused. And in practice, at least in my limited experience, ctrl-P from the score tab is faster and more to the point when wanting to print the score. 

In short, I think the Print dialog needs to look like it was actually implemented (I'm looking at you, "Printing from Win32 Application. - Print") in Musescore, one of the biggest names in Music.. err... scores ;-) 

@cbjeukendrup you highlight an important potential blocker to this issue. If QT could be fixed it will help a bunch of other projects using it - that has to be the ideal in principle. Writing musescore code for each OS version sounds like a right pain. I hope the QT team are up for it ;-)

**Comment by MarcSabatella** [comment]
What I don't understand is why you'd bother opening a print dialog if you just want to see a preview.  In what way is the Publish tab not exactly that but better?  If you want to actually print, sure, use the dialog, but to see the preview, no need - just look at the Publish tab.  That *is* a print preview.  if you want to see a preview and then print when satisfied, click Publish then go to the dialog when satisfied.

I'm not saying it wouldn't also be nice if Qt were to implement their own print preview mode, but meanwhile, there is nothing I can think of that it would do that you can't do just as easily today.

**Comment by GabeS573** [comment]
It's hard to call the Publish tab an appropriate "Print Preview" if it doesn't always preview what you're about to print. If you have a landscape score, go to the Publish tab, and hit Print without making any changes, it prints in portrait and cuts off the sides of every single page. Ink and paper aren't free...

The whole point of having a print preview built into the print dialog is so that you can verify your document is going to print exactly the way you want it to. The publish tab is not a print preview, and it won't be unless it can take into account options like paper size (independent of the page size saved in the mscx), page orientation (same thing), printing multiple pages per sheet, grayscale printing, paper finish, and all the other printer-specific options that a print dialog has. Imo, it would be rather unconventional if the Publish tab had all of this built into it, which is why I think the print dialog should be fixed.

> What I don't understand is why you'd bother opening a print dialog if you just want to see a preview.

Nobody opens the print dialog "just" to see a preview, we open the print dialog when we want to... print. Seeing a dynamic preview of *exactly* what is going to come out of our printer tray is an important part of that.

**Comment by MarcSabatella** [comment]
OK, I hadn't considered the issue of wrong orientation or paper size.  In theory, those do get set automatically in the print dialog based on the actual score properties so you shouldn't need to change any settings anyhow, but it probably does not work reliably for all operating systems or all printer drivers.  And for 4-up printing or other more esoteric options, Publish obviously won't show that.

Of course, even a native print dialog can't show you if you have the wrong size paper actually loaded, so there is no 100% foolproof way to know exactly what you will get.

Anyhow, the point remains - this isn't something that is easily fixed because the print dialogs are not part of MuseScore's code, so I'm simply trying to explain the options that *do* exist in order to help people see how to solve problems *today*.

The Publish tab does present most of what a print preview would normally be used for, and I think many people don't realize that, so it's still useful info.  For those cases where you are doing something out of the ordinary and you feel a more accurate preview would be useful, you can always export to PDF then print that normally to see your system's native print dialog.

Again, this isn't to say it wouldn't be nice if someday Qt implemented a print preview, but these are the workarounds for now.  Publish for the normal cases, PDF for the corner cases.

**Comment by juli27** [comment]
I've prototyped a `WindowsPrintProvider` which uses the modern print API directly. ([Downloadable build to try it out](https://github.com/juli27/MuseScore/actions/runs/19213119151), [code](https://github.com/musescore/MuseScore/compare/master...juli27:MuseScore:modernizePrintingWindows))

All  points are implemented except:
> - If you choose to print a custom page range, the input isn't constrained to pages in the document (although only pages actually in the document will print)

This is impossible to do with the new print dialog as far as I can tell. Is there any app where it does work with the modern dialog?

My prototype has the following shortcomings:
- Not all paper sizes are implemented
- Both the preview and final output are blurry (probably some DPI issue that needs figuring out)
- *And most importantly:* The data that is sent to the driver is no longer vector data, but pixel data instead. This causes a loss of quality & data size explosion (especially when printing to a file). In order to keep sending vector data, we would have to implement a completely separate painting backend based on either the XPS-document API (this is the preferred format for printers on Windows), Direct2D (code could be reused for HW-accelerated painting on Windows only), or some third-party library with a compatible backend (like [Skia](https://skia.org); could be used for HW-accelerated painting on all platforms)

Unless it's already planned that we use a different backend for painting, I don't think this his is worth pursuing for this issue alone. So for me, that last item is reason enough to not go forward with this approach.

An alternative approach would be to implement our own, platform independent, print dialog (Mozilla Firefox, Google Chrome, Microsoft Office, LibreOffice all have one). This way, we could use the `QPrinter` API and therefore keep using the existing painting code. 
I see no other way, other than leaving it as-is.

**Comment by GabeS573** [comment]
Oof, that's unfortunate. Needless to say the quality of the actual print is more important than the dialog being completely functional. It sounds like this issue is not going to be fixed without going the custom dialog route.

**Comment by cbjeukendrup** [comment]
Thanks for the investigation, @juli27! Yeah, in theory it should be possible to create a `IPaintProvider` implementation to use here, but that's more work and more risk than we can afford.

Actually, I do believe there are plans to turn Publish mode in a more useful print preview, also with printing settings right there, so that it would essentially be the same as a custom printing dialog. (@jessjwilliamson I believe you worked on designs for this, is that right?)

**Comment by jessjwilliamson** [comment]
> Thanks for the investigation, [@juli27](https://github.com/juli27)! Yeah, in theory it should be possible to create a `IPaintProvider` implementation to use here, but that's more work and more risk than we can afford.
> 
> Actually, I do believe there are plans to turn Publish mode in a more useful print preview, also with printing settings right there, so that it would essentially be the same as a custom printing dialog. ([@jessjwilliamson](https://github.com/jessjwilliamson) I believe you worked on designs for this, is that right?)

Yes, there is a design project about having a printing functionality within the publish tab. It's not finished and I don't know when we'll bring this back up again

**Comment by Jojo-Schmitz** [comment]
Came up again in https://musescore.org/en/node/394660

**Comment by polomora** [comment]
I was redirected to this bug report from the Musescore bug reporting forum.
I've discovered another issue. Copied from that forum bug report:

> I usually print my scores single-sided, ie when I print a score I turn off the double-sided print option,
> using Control + P -> More Settings -> 2-sided / Booklet = None.
> The print dialog that appears on the screen is "Printing from Win32 application - Print".
> 
> I notice that after setting this option to None, all other print requests from all other apps are also printed with the Duplex print option turned off. In the Printer Properties for this printer, accessed using Windows + I, the default value for the 2-sided / Booklet was changed to None. That is, Muscscore modified the default duplex option for the printer.

Thank you.
